src/sc_stage.py

```diff
@@ -303,56 +322,104 @@ def mark_item_delivered(self, message: Message):
+        log.debug("Node %d sending a material batch request for item type %s to node %d", self.node_id, supplier_type, supplier_id)
+
+        batch_request = BatchRequest(
+            source=self.node_id,
+            dest=supplier_id,
+            item_req=ItemReq(Item(supplier_type, None), 1),  # HACK only batches of size 1 requested
+            request_id=str(uuid4())[:8],  # HACK for unique but short request IDs
+        )
+        self.send_message(batch_request)
+        self.pending_requests[batch_request.request_id] = batch_request
+
+        log.debug("Node %d sent a material batch request %s", self.node_id, batch_request)
 
-            This is a blocking call since it has to wait on the inbound queue.
+    def _attempt_manufacture_cycle(self, suppliers):
         '''
-        if self.item_dep.has_prereq():
-            try:
-                # For each manufactured batch item, do a get() in the inbound queue once.
-                # NOTE/FIXME/TODO
-                # This does not work if batches of different item quantities
-                # and item types are required by the stage's inbound requirements.
-                # The manufacture logic assumes each result batch can be made with
-                # any, single, incoming batch.
-                log.info("Node %d waiting for %s from %s", self.node_id, self.item_dep.get_prereq(), suppliers)
-                material_req = self.inbound_material.get(timeout=5)  # HACK wait at most 5 seconds for a material batch
-                self.inbound_log[material_req] = StageStatus.CONSUMED
-                log.info("Node %d successfully consumed %s", self.node_id, material_req)
-            except Empty:
-                # queue is empty, need to wait till it's not.
-                log.warning("Node %d's inbound material queue is empty but needs %s. Skipping manufacturing of one batch of item type %s",
-                            self.node_id, self.item_dep.get_prereq(), self.get_stage_result_type())
+            suppliers: list of (node_id, item_req) tuples obtained from the latest flow
+        '''
+        prereqs = self.item_dep.get_prereq()
+        if prereqs:
+            queue_types = set(self.inbound_material.keys())
+            supplier_types = set([ir.item.type for _, ir in suppliers])
+
+            suppliers = {ir.item.type: sid for sid, ir in suppliers}  # convert the flow edges to dict
+
+            if not queue_types.issuperset(supplier_types):  # verify there is queue for each supplier
+                log.critical("Node {} has more suppliers ({}) from flow the available queue types ({})"
+                             .format(self.node_id, supplier_types, queue_types))
+                log.error("Node %d skipping manufacturing cycle", self.node_id)
+                return
+
+            has_all_materials = True
+            for item_type, supplier_id in suppliers.items():
+                queue = self.inbound_material[item_type]
+
+                if queue.empty():
```

Comment:
Where do we handle different quantities of materials? What if a node requires K number of items.

Reply:
another assumption. batch sizes are always 1.